backend/rest/rest.go

```diff
@@ -0,0 +1,215 @@
+package rest
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/restic/restic/backend"
+)
+
+const connLimit = 10
+
+// Returns the url of the resource
+func restPath(url *url.URL, t backend.Type, name string) string {
+	location := url.String()
+	if !strings.HasSuffix(location, "/") {
+		location += "/"
+	}
+	if t == backend.Config {
+		return location + string(t)
+	}
+	return location + string(t) + "/" + name
+}
+
+type RestBlob struct {
+	b     *Rest
+	buf   *bytes.Buffer
+	final bool
+}
+
+func (rb *RestBlob) Write(p []byte) (int, error) {
+	if rb.final {
+		return 0, errors.New("blob already closed")
+	}
+	n, err := rb.buf.Write(p)
+	return n, err
+}
+
+func (rb *RestBlob) Read(p []byte) (int, error) {
+	return rb.buf.Read(p)
+}
+
+func (rb *RestBlob) Close() error {
+	rb.final = true
+	rb.buf.Reset()
+	return nil
+}
+
+// Finalize moves the data blob to the final location for type and name.
+func (rb *RestBlob) Finalize(t backend.Type, name string) error {
+	if rb.final {
+		return errors.New("already finalized")
+	}
+
+	rb.final = true
+
+	// Check key does not already exist.
```

Comment:
I agree. This saves a round-trip for each newly added blob. Because the error condition we're trying to avoid here (uploading a duplicate blob) is unlikely, this should speed up the regular case (new blob) greatly.